File: src/Commands/Docs.fs
/// <summary>
/// Provides /docs and /source commands to search the CC:T and Lua docs for a
/// particular method.
///
/// We also provide auto-completion for these methods sources, as well as legacy
/// %docs and %source commands.
/// <summary>
module FAQBotCC.Commands.Docs

open System.Collections.Generic
open System.Globalization
open System.Net.Http
open System.Text.Json
open System.Text.Json.Serialization
open System.Threading.Tasks

open Discord
open Discord.Interactions
open Discord.Commands

open Microsoft.Extensions.Logging

open FAQBotCC
open FAQBotCC.Extensions


type private ArgSummaryAttribute = Interactions.SummaryAttribute


/// <summary>A documented item inside the illuaminate doc export</summary>
type DocumentedItem =
  { [<JsonPropertyName("name")>]
    Name : string
    [<JsonPropertyName("source")>]
    Source : string
    [<JsonPropertyName("summary")>]
    Summary : Option<string>
    [<JsonPropertyName("module")>]
    Module : string
    [<JsonPropertyName("module-kind")>]
    ModuleKind : string
    [<JsonPropertyName("url")>]
    Url : string }


module DocumentedItem =
  let makeEmbed (item : DocumentedItem) link =
    let embed = EmbedBuilder()
    embed.Title <- item.Name
    embed.Url <- link item
    match item.Summary with
    | None -> ()
    | Some summary -> embed.Description <- summary
    embed.Build()

  let linkDocs (item : DocumentedItem) = $"https://tweaked.cc/{item.Url}"
  let linkSource (item : DocumentedItem) = item.Source


/// <summary>Wrapper around a CachedRequest which is used with DI.</summary>
type MethodStore(logger : ILogger<MethodStore>, client : HttpClient) =
  let cache : CachedRequest.t<IReadOnlyDictionary<_, _>> =
    CachedRequest.make client logger "https://tweaked.cc/index.json" (int64 (60 * 1000)) (fun x ->
      JsonSerializer.Deserialize<Dictionary<string, DocumentedItem>>(x))

  member _.Get = CachedRequest.get cache


/// <summary>
/// Provides autocompletion for a method name.
///
/// For now, this just works off a .StartsWith check, and scores based on length.
/// </summary>
type MethodNameAutocompleteHandler(methods : MethodStore) =
  inherit AutocompleteHandler()

  override _.GenerateSuggestionsAsync
    (
      _context : IInteractionContext,
      autocompleteInteraction : IAutocompleteInteraction,
      _parameter : IParameterInfo,
      _services : System.IServiceProvider
    ) : Task<AutocompletionResult> =
    task {
      let value = autocompleteInteraction.Data.Current.Value.ToString().ToLowerInvariant()
      let! methods = methods.Get
      return
        Seq.append methods.Keys LuaNames.vars.Keys
        |> Seq.filter (fun name -> name.StartsWith(value, ignoreCase = true, culture = CultureInfo.InvariantCulture))
        |> Seq.sortBy (fun x -> x.Length)
        |> Seq.truncate 25
        |> Seq.map (fun name -> AutocompleteResult(name, name))
        |> AutocompletionResult.FromSuccess
    }


let private run (context : IDiscordContext) (methods : MethodStore) (query : string) link : Task =
  task {
    let! methods = methods.Get
    let query = query.TrimEnd('(', ')', ' ')
    // First attempt to find a CC:T method.
    match Lookup.find methods query with
    | Lookup.Exact value -> return! context.Respond(embed = DocumentedItem.makeEmbed value link)
    | Lookup.Fuzzy (name, item) ->
      return!
        context.Respond(
          text = $"Cannot find '{query}', using '{name}' instead.",
          embed = DocumentedItem.makeEmbed item link
        )
    | Lookup.Missing ->
      // Failing that, find a built-in Lua one.
      let luaQuery = query.ToLowerInvariant()
      match Lookup.find LuaNames.vars luaQuery with
      | Lookup.Exact link -> return! context.Respond(embed = EmbedBuilder().WithTitle(luaQuery).WithUrl(link).Build())
      | Lookup.Fuzzy (name, link) ->
        return!
          context.Respond(
            text = $"Cannot find '{query}', using '{name}' instead.",
            embed = EmbedBuilder().WithTitle(name).WithUrl(link).Build()
          )
      | Lookup.Missing ->
        return!
          context.Respond(
            text =
              $"Cannot find '{query}'. Please check your spelling, or contribute to the documentation at https://github.com/cc-tweaked/CC-Tweaked."
          )
  }


module private Descriptions =
  [<Literal>]
  let arg = "The function's name"

  [<Literal>]
  let docs =
    "Searches for a function with the current name and returns its documentation."

  [<Literal>]
  let source =
    "Searches for a function with the current name, and returns a link to its source code."


type DocsTextCommand(methods : MethodStore) =
  inherit ModuleBase<ICommandContext>()

  [<Command("docs")>]
  [<Alias("d")>]
  [<Summary(Descriptions.docs)>]
  member this.Docs([<Summary(Descriptions.arg)>] name : string) =
    run this.Context.DiscordContext methods name DocumentedItem.linkDocs

  [<Command("source")>]
  [<Alias("s")>]
  [<Summary(Descriptions.source)>]
  member this.Source([<Summary(Descriptions.arg)>] name : string) =
    run this.Context.DiscordContext methods name DocumentedItem.linkSource


type DocsInteractionCommand(methods : MethodStore) =
  inherit InteractionModuleBase<IInteractionContext>()

  [<SlashCommand("docs", Descriptions.docs)>]
  member this.Docs
    ([<ArgSummary(description = Descriptions.arg); Autocomplete(typedefof<MethodNameAutocompleteHandler>)>] name : string)
    =
    run this.Context.DiscordContext methods name DocumentedItem.linkDocs

  [<SlashCommand("source", Descriptions.source)>]
  member this.Source
    ([<ArgSummary(description = Descriptions.arg); Autocomplete(typedefof<MethodNameAutocompleteHandler>)>] name : string)
    =
    run this.Context.DiscordContext methods name DocumentedItem.linkSource

File: src/commands/docs.rs
//! Provides `/docs` and `/source` commands to search the CC:T and Lua docs fo
//! a particular method.

use std::{borrow::Borrow, collections::HashMap, hash::Hash};

use fuzzy_matcher::FuzzyMatcher;
use serenity::{
  all::prelude::Context,
  builder::{
    CreateAutocompleteResponse, CreateCommand, CreateCommandOption, CreateEmbed,
    CreateInteractionResponse, CreateInteractionResponseMessage,
  },
  model::application::{
    CommandInteraction, CommandOptionType, CommandType, ResolvedOption, ResolvedValue,
  },
};

use crate::{InteractionError, State, state::DocumentedEntry};

enum EmbedMode {
  Docs,
  Source,
}

async fn lookup(
  state: &State,
  name: &str,
  mode: EmbedMode,
) -> Result<CreateInteractionResponseMessage, InteractionError> {
  let name = name.trim_end_matches(['(', ')', ' ']);

  let methods = match state.documentation_entries().await {
    Ok(x) => x,
    Err(err) => {
      tracing::error!(?err, "Cannot get CC:T definitions");
      return Ok(
        CreateInteractionResponseMessage::new()
          .content("Failed to fetch CC:T definitions")
          .ephemeral(true),
      );
    }
  };

  match fuzzy_lookup(&methods, name) {
    FuzzyLookup::Exact(d) => Ok(CreateInteractionResponseMessage::new().embed(make_embed(d, mode))),
    FuzzyLookup::Fuzzy(d) => Ok(CreateInteractionResponseMessage::new()
        .content(format!("Cannot find '{}', using {}' instead", name, &d.name))
        .embed(make_embed(d, mode))
      ),
    FuzzyLookup::Missing => {
      Ok(CreateInteractionResponseMessage::new()
          .content(format!(
            "Cannot find '{name}'. Please check your spelling, or contribute to the documentation at https://github.com/cc-tweaked/CC-Tweaked."
          ))
        .ephemeral(true)
      )
    }
  }
}

async fn run(
  state: &State,
  ctx: Context,
  interaction: CommandInteraction,
  mode: EmbedMode,
) -> Result<(), InteractionError> {
  let &[
    ResolvedOption {
      name: "name",
      value: ResolvedValue::String(search),
      ..
    },
  ] = interaction.data.options().as_slice()
  else {
    return Err(InteractionError::InvalidOptions);
  };

  interaction
    .create_response(
      &ctx,
      CreateInteractionResponse::Message(lookup(state, search, mode).await?),
    )
    .await?;
  Ok(())
}

/// A basic auto-complete handler for a method name.
///
/// This just finds all strings prefixed with the current text, then scores
/// based on length.
pub async fn autocomplete(state: &State, search: &str) -> CreateAutocompleteResponse {
  let search = search.trim();
  let methods = match state.documentation_entries().await {
    Ok(x) => x,
    Err(err) => {
      tracing::error!(?err, "Cannot get CC:T definitions");
      return CreateAutocompleteResponse::new();
    }
  };

  let mut choices: Vec<_> = methods
    .keys()
    .filter(|x| x.len() >= search.len() && x[..search.len()].eq_ignore_ascii_case(search))
    .collect();
  choices.sort_by_key(|x| x.len());

  choices
    .into_iter()
    .take(25)
    .fold(CreateAutocompleteResponse::new(), |r, x| {
      r.add_string_choice(x, x)
    })
}

pub const DOCS_ID: &str = "docs";
pub const SOURCE_ID: &str = "source";
pub const METHOD_ARG: &str = "name";

pub fn register_docs() -> CreateCommand {
  CreateCommand::new(DOCS_ID)
    .kind(CommandType::ChatInput)
    .description("Searches for a function with the current name and returns its documentation.")
    .add_option(
      CreateCommandOption::new(CommandOptionType::String, METHOD_ARG, "The function's name")
        .required(true)
        .set_autocomplete(true),
    )
}

/// Searches for a function with the current name and returns its documentation.
pub async fn run_docs(
  state: &State,
  ctx: Context,
  interaction: CommandInteraction,
) -> Result<(), InteractionError> {
  run(state, ctx, interaction, EmbedMode::Docs).await
}

pub fn register_source() -> CreateCommand {
  CreateCommand::new(SOURCE_ID)
    .kind(CommandType::ChatInput)
    .description(
      "earches for a function with the current name and returns a link to the source code.",
    )
    .add_option(
      CreateCommandOption::new(CommandOptionType::String, METHOD_ARG, "The function's name")
        .required(true)
        .set_autocomplete(true),
    )
}

/// Searches for a function with the current name and returns a link to the source code.
pub async fn run_source(
  state: &State,
  ctx: Context,
  interaction: CommandInteraction,
) -> Result<(), InteractionError> {
  run(state, ctx, interaction, EmbedMode::Source).await
}

enum FuzzyLookup<T> {
  Missing,
  Fuzzy(T),
  Exact(T),
}

/// Look up an entry in a map, either returning an exact match ([FuzzyLookup::Exact]) or a
/// partial/fuzzy one ([FuzzyLookup::Fuzzy]).
fn fuzzy_lookup<'a, K, V>(map: &'a HashMap<K, V>, pattern: &'a str) -> FuzzyLookup<&'a V>
where
  K: Eq + Hash + Borrow<str>,
{
  if let Some(x) = map.get(pattern) {
    return FuzzyLookup::Exact(x);
  }

  let matcher = fuzzy_matcher::skim::SkimMatcherV2::default();
  let mut matches: Vec<_> = map
    .iter()
    .filter_map(|(k, v)| matcher.fuzzy_match(k.borrow(), pattern).map(|s| (v, s)))
    .collect();
  matches.sort_by_key(|(_, x)| *x);

  match *matches.as_slice() {
    [(x, _)] => FuzzyLookup::Fuzzy(x),
    [.., (x, x_score), (_, y_score)] if x_score > y_score + 5 => FuzzyLookup::Fuzzy(x),
    _ => FuzzyLookup::Missing,
  }
}

fn make_embed(entry: &DocumentedEntry, mode: EmbedMode) -> CreateEmbed {
  let mut embed = CreateEmbed::new()
    .title(entry.name.as_str())
    .url(match mode {
      EmbedMode::Docs => &entry.docs,
      EmbedMode::Source => entry.source.as_ref().unwrap_or(&entry.docs),
    });

  if let Some(summary) = entry.summary.as_ref() {
    embed = embed.description(summary);
  }

  embed
}
